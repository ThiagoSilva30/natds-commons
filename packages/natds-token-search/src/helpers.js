import {
  pipe, unless, map, objOf, zipObj, mapObjIndexed, isNil,
  filter, prop, contains, toLower, anyPass, toString,
} from 'ramda';

export const flattenData = (properties) => {
  const mapProps = (item) => Object.values(item).map(checkIfHasName); // eslint-disable-line
  const checkIfHasName = (item) => (item.name ? item : mapProps(item));

  return mapProps(properties).flat(2);
};

export const buildTokensForBrand = (brand) => pipe(
  map(flattenData),
  zipObj(['light', 'dark']),
  objOf(brand),
);

const setAttribute = (element) => (attribute, key) => element.setAttribute(key, attribute);
const setContent = (element) => (content) => {
  element.innerHTML = content; // eslint-disable-line
};

export const createElement = (element, attributes, content) => {
  const el = document.createElement(element);

  unless(
    isNil,
    mapObjIndexed(setAttribute(el)),
  )(attributes);

  unless(
    isNil,
    setContent(el),
  )(content);

  return el;
};

const searchRules = (search) => anyPass([
  pipe(prop('name'), toLower, contains(toLower(search))),
  pipe(prop('value'), toString, toLower, contains(toLower(search))),
]);

export const filterTokens = (tokens, search) => {
  if (!search) return tokens;

  return filter(searchRules(search), tokens);
};
