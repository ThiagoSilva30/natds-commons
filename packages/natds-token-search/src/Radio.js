import { createElement } from './helpers';

export const Radio = (name, value, checked) => {
  const label = createElement('label', { for: value }, value);

  const input = createElement('input', {
    id: value,
    name,
    type: 'radio',
    value,
    ...(checked ? { checked: true } : {}),
  });

  const wrapper = createElement('span');

  wrapper.appendChild(input);
  wrapper.appendChild(label);

  return wrapper;
};

export const RadioGroup = (name, data, selected) => {
  const wrapper = createElement('div');

  wrapper.classList.add(name);

  data.forEach((item) => {
    wrapper.appendChild(Radio(name, item, selected === item));
  });

  return wrapper;
};
