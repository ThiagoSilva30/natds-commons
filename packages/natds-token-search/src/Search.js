import { createElement } from './helpers';
import jss from './styles/jss';

const inputStyles = {
  input: {
    border: [1, 'solid', '#DFDFDF'],
    margin: [[30, 0]],
    padding: 8,
    width: 480,
  },
};

const sheet = jss.createStyleSheet(inputStyles);

export const Search = (onInput) => {
  sheet.attach();
  const input = createElement('input', { type: 'search' }, null, sheet.classes.input);

  input.addEventListener('input', onInput);

  return input;
};

export default Search;
