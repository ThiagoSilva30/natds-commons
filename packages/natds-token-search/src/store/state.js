import { createStore } from 'redux';
import { path, prop, mergeDeepRight } from 'ramda';
import { tokens as allTokens } from './tokens';
import { searchActionType } from './actions';
import { filterTokens } from '../helpers';

const initialState = {
  allTokens,
  brand: 'natura',
  mode: 'light',
  platform: 'web',
  search: null,
  selectedTokens: allTokens.web.natura.light,
};

const onSearch = (state, { payload }) => {
  const newState = mergeDeepRight(state, payload);
  const { platform, brand, mode } = newState;

  const selectedTokens = filterTokens(
    path([platform, brand, mode], state.allTokens),
    newState.search,
  );

  return {
    ...newState,
    selectedTokens,
  };
};

const reducer = (state = initialState, action) => {
  const actionMap = {
    [searchActionType]: onSearch,
  };

  const currentAction = prop(action.type, actionMap);

  console.log({ state });
  console.log({ action });

  return currentAction ? currentAction(state, action) : { ...state };
};

export const store = createStore(reducer);

export default store;
