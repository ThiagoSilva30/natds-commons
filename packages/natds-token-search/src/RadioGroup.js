import { createElement } from './helpers';
import jss from './styles/jss';
import { Radio } from './Radio';

const radioGroupStyles = {
  radios: {
    '& span': {
      '& *': {
        verticalAlign: 'middle',
      },
      '& input': {
        margin: 0,
      },
      '& label': {
        fontSize: 14,
        paddingLeft: 8,
      },
      padding: [8, 0],
      width: 120,
    },
    display: 'flex',
  },
  radioGroup: {
    '&.platform': {
      '& $radios': {
        '&::after': {
          borderBottom: [1, 'solid', '#dfdfdf'],
          bottom: 0,
          content: '""',
          left: 16,
          position: 'absolute',
          right: 0,
        },
        marginLeft: -20,
        position: 'relative',
      },
      '& h5': {
        display: 'none',
      },
      '& input': {
        '&:checked+label': {
          fontWeight: 'bold',
        },
        '&:checked+label::after': {
          borderBottom: [2, 'solid', '#227BBD'],
          bottom: 0,
          content: '""',
          left: 0,
          position: 'absolute',
          right: 0,
        },
        opacity: 0,
        position: 'absolute',
      },
      '& label': {
        padding: [20, 0],
        position: 'relative',
      },
      '& span': {
        display: 'flex',
        margin: [0, 16],
        padding: 0,
        width: 'auto',
      },
    },
    margin: [20, 0],
  },
};

const radioGroupSheet = jss.createStyleSheet(radioGroupStyles);

export const RadioGroup = (name, data, selected, titleText) => {
  radioGroupSheet.attach();
  const wrapper = createElement('div', null, null, radioGroupSheet.classes.radioGroup);
  const title = createElement('h5', null, titleText);

  wrapper.classList.add(name);
  wrapper.appendChild(title);

  const radios = createElement('div', null, null, radioGroupSheet.classes.radios);

  data.forEach((item) => {
    radios.appendChild(Radio(name, item, selected === item));
  });

  wrapper.appendChild(radios);

  return wrapper;
};

export default RadioGroup;
