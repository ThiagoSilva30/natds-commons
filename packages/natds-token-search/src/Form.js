import { createElement } from './helpers';
import { searchAction, store } from './store';
import { RadioGroup } from './Radio';
import { Search } from './Search';

const brands = [
  'aesop',
  'avon',
  'natura',
  'theBodyShop',
];

const modes = ['light', 'dark'];

const platforms = [
  'android',
  'ios',
  'reactNative',
  'web',
];

const onInputSearch = ({ target }) => store.dispatch(
  searchAction({ search: target.value }),
);

const onFormChange = ({ target }) => store.dispatch(
  searchAction({ [target.name]: target.value }),
);

export const Form = () => {
  const form = createElement('form');

  const state = store.getState();

  form.appendChild(RadioGroup('platform', platforms, state.platform, 'Platforms'));
  form.appendChild(Search(onInputSearch));
  form.appendChild(RadioGroup('brand', brands, state.brand, 'Brands'));
  form.appendChild(RadioGroup('mode', modes, state.mode, 'Modes'));

  form.addEventListener('change', onFormChange);

  return form;
};

export default Form;
