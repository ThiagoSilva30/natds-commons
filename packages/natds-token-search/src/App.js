import { Table } from './Table';
import { Form } from './Form';
import { createElement } from './helpers';

export const App = () => {
  const wrapper = createElement('div');

  wrapper.appendChild(Form());
  wrapper.appendChild(Table());

  return wrapper;
};

export default App;
