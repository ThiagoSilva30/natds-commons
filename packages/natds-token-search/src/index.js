import { App } from './App';

if (module.hot) module.hot.accept();

const root = document.getElementById('app-root');

root.appendChild(App());
