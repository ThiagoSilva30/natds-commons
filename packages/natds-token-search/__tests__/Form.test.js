import { getByText, queryByTestId, getByTestId, fireEvent } from '@testing-library/dom';
import { Form } from '../src/Form';
import * as Search from '../src/Search';
import * as RadioGroup from '../src/RadioGroup';
import * as helpers from '../src/helpers';
import { store } from '../src/store';
import * as actions from '../src/store/actions';
import { createMockElement } from './testHelpers';

const createMockInput = (type, name) => createMockElement('input', {
  type,
  name,
  'data-testid': name,
});


const state = {
  allTokens: 'allTokens',
  brand: 'natura',
  mode: 'light',
  platform: 'web',
  search: null,
  selectedTokens: 'selectedTokens',
};

describe('Form', () => {
  let createElementSpy;
  let radioGroupSpy;

  beforeEach(() => {
    jest.spyOn(Search, 'Search').mockReturnValue(createMockInput('search', 'search'));
    jest.spyOn(store, 'getState').mockReturnValue(state);
    radioGroupSpy = jest.spyOn(RadioGroup, 'RadioGroup').mockImplementation((name) => createMockInput('radio', name));
    createElementSpy = jest.spyOn(helpers, 'createElement');
  });

  it('calls createElement', () => {
    Form();

    expect(createElementSpy).toHaveBeenCalled();
  });

  it('should render the form', () => expect(Form()).toMatchSnapshot());

  it('should call search action on form change', () => {
    const searchActionSpy = jest.spyOn(actions, 'searchAction');

    const form = Form();
    fireEvent.change(getByTestId(form, 'platform'), { target: { value: 'android', name: 'platform' } })

    expect(searchActionSpy).toHaveBeenCalledWith({ platform: 'android' })
  });

  describe('radio group', () => {
    it('should render the form platform radio group', () => {
      const form = Form();

      expect(queryByTestId(form, 'platform')).toBeTruthy()
      expect(radioGroupSpy.mock.calls[0]).toEqual(['platform', ['android', 'ios', 'reactNative', 'web'], 'web', 'Platforms'])
    });

    it('should render the form brand radio group', () => {
      const form = Form();

      expect(queryByTestId(form, 'brand')).toBeTruthy()
      expect(radioGroupSpy.mock.calls[1]).toEqual(['brand', ['aesop', 'avon', 'natura', 'theBodyShop'], 'natura', 'Brands'])
    });

    it('should render the form mode radio group', () => {
      const form = Form();

      expect(queryByTestId(form, 'mode')).toBeTruthy()
      expect(radioGroupSpy.mock.calls[2]).toEqual(['mode', ['light', 'dark'], 'light', 'Modes'])
    });
  });

  describe('search', () => {
    it('should render the form search', () => expect(queryByTestId(Form(), 'search')).toBeTruthy());

    it('should call search action on search input', () => {
      const searchActionSpy = jest.spyOn(actions, 'searchAction');

      const form = Form();

      fireEvent.change(getByTestId(form, 'search'), { target: { value: 'pikachu' } })

      expect(searchActionSpy).toHaveBeenCalledWith({ search: 'pikachu' })
    });
  });
});
