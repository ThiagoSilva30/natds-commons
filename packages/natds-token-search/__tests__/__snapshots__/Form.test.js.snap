// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render the form 1`] = `
<form>
  <input
    data-testid="platform"
    name="platform"
    type="radio"
  />
  <input
    data-testid="search"
    name="search"
    type="search"
  />
  <input
    data-testid="brand"
    name="brand"
    type="radio"
  />
  <input
    data-testid="mode"
    name="mode"
    type="radio"
  />
</form>
`;
